### ISSUE TITLE
Changes to Rancher internal DNS

### ISSUE THREAD

**Issue by alena1108** [rancher/rancher#2628]
To be more aligned with docker 1.9 networking model, we should do the following for our Rancher internal DNS:

- service is discoverable to every other service within the same stack by default. Service links would only be useful if user wants to override DNS name
- every container is discoverable by its name to every other container in the same stack
- to make service discoverable to outside of the stack, links should still be used. 
- for service to register to the LB, links still be used.

@ibuildthecloud @will-chan

**Comment by sangeethah** [comment]
Tested with server version - v0.46.0-rc2

1.Services within the same stack are discoverable to each other using their service names.
2.Containres within the same stack are discoverable to each other using their container_names.

**Comment by alex88** [comment]
After upgrading to this version in both rancher and compute hosts with the agent, nginx frontends fails to start due

```
[emerg] 1#1: host not found in upstream "api:8000" in /etc/nginx/nginx.conf:38
[emerg] host not found in upstream "api:8000" in /etc/nginx/nginx.conf:38
```

with the host linked, I've to check how I can use service names instead since we've multiple environments

**Comment by alex88** [comment]
Tried also with service name (they're in the same stack), still nginx error, using ip address everything works

**Comment by pedrommone** [comment]
Same happens to me. If I force the `fpm` IP it works well, if I use a DNS, it falls.

**Comment by alex88** [comment]
@pedrommone are you using nginx? how's your config?

**Comment by pedrommone** [comment]
@alex88 yep!

```
user www-data;
worker_processes 4;
pid /run/nginx.pid;
daemon off;

events {
    worker_connections 768;
    # multi_accept on;
}

http {
    upstream phpfpm_backend {
        server %fpm-ip%:9000;
    }

    ##
    # Basic Settings
    ##

    sendfile off;
    tcp_nopush on;
    tcp_nodelay on;
    keepalive_timeout 65;
    types_hash_max_size 2048;

    # server_names_hash_bucket_size 64;
    # server_name_in_redirect off;

    include /etc/nginx/mime.types;
    default_type application/octet-stream;

    ##
    # Logging Settings
    ##

    access_log /data/logs/access.log;
    error_log /data/logs/error.log warn;

    ##
    # Gzip Settings
    ##

    gzip on;
    gzip_disable "msie6";

    include /etc/nginx/conf.d/*.conf;
}
```

**Comment by alex88** [comment]
@pedrommone how do you replace that %fpm-ip%? Is it a linked docker service?

**Comment by pedrommone** [comment]
it's a linked docker. I replace with `fpm`

heres the docker-composer.yml
```
atena:
    image: registry.pedro.com:443/atena
    tty: true
    stdin_open: true
    labels:
        io.rancher.container.pull_image: always
    command: cat
fpm:
    image: php:7-fpm
    tty: true
    stdin_open: true
    labels:
        io.rancher.sidekicks: atena
        io.rancher.scheduler.affinity:host_label: server=08
    volumes_from:
        - atena
nginx:
    image: registry.pedro.com:443/laravel-nginx
    links:
        - fpm:fpm
    labels:
        io.rancher.sidekicks: atena
        io.rancher.scheduler.affinity:host_label: server=08
    volumes_from:
        - atena
postgres:
    image: postgres
    volumes:
        - /var/lib/postgresql/atena:/var/lib/postgresql/data
    environment:
        POSTGRES_USER: atena
        POSTGRES_PASSWORD: deuspedrolindo
    labels:
        io.rancher.scheduler.affinity:host_label: server=08
```

**Comment by alex88** [comment]
First of all, I think you can just use `links: - fpm` without `fpm:fpm`, it's redundant
Second, that nginx config will resolve the fpm dns only on boot, probably even before rancher has a response for that, so you'll get that error. I solved the issue with:

```
    location /api {
      resolver 169.254.169.250 valid=5s ipv6=off;
      set $api "api";
      proxy_pass http://$api:8000;
    }
```

This will resolve the dns every 5 seconds it needs to connect to `api`. Be sure to set `ipv6=off` otherwise nginx won't correctly resolve the address.

**Comment by pedrommone** [comment]
Worked after remove `upstream` from nginx and used this block of code

```
location ~ \.php$ {
    resolver 169.254.169.250 valid=5s ipv6=off;
    include fastcgi_params;
    set $fpm "fpm";
    fastcgi_pass $fpm:9000;
    fastcgi_param SCRIPT_FILENAME $request_filename;
}
```

Pro tip: do not use upstream.

**Comment by akurkin** [comment]
hey @alex88 and @pedrommone 
having the same problem and I always used upstream, but now looking how you guys solved this. Can you please advise what is this ip `169.254.169.250` for resolver?

**Comment by pedrommone** [comment]
@akurkin did you follow our config? Thats the way I've solved it.

**Comment by akurkin** [comment]
@pedrommone yes, I did! worked for me too, thanks much!

**Comment by alex88** [comment]
@akurkin that's the IP of the ranche doccker container running on each server, where the ranche dns knows all the container names and so can resolve the correct ips

**Comment by looztra** [comment]
Same problem for me (and same working fix) with latest version v0.63.0

**Comment by alex88** [comment]
I won't call if a bug, probably it takes a few seconds to the name to resolve correctly, the thing is the nginx configuration is not so straightforward
